crates/bevy_input_focus/src/directional_navigation.rs

```diff
@@ -184,6 +197,168 @@ impl DirectionalNavigationMap {
+    /// in the [`SouthEast`](CompassOctant::SouthEast) direction, but would never see an entity at (1, 2).
+    /// That said, you can add the same entity at multiple locations in the grid!
+    /// This works well to model larger entities,
+    /// even if the entities aren't truly grid-aligned.
+    ///
+    /// If `should_loop` is set to `true`, the top and bottom rows will be connected, as will the left and right columns.
+    // The keys are specified as i32 for compatibility with taffy's grid system,
+    // which uses u16 values. We need signed integers here, but we should be able to cast them losslessly.
+    // PERF: This is a very flexible / "clever" implementation, but it won't be the fastest for simple cases.
+    // If there's user demand for it, we can add a more optimized / less flexible version that requires non-sparse rectangular grids.
+    pub fn add_grid(&mut self, entity_grid: NavGrid, should_loop: bool) {
+        let mut min_rows = i32::MAX;
+        let mut min_columns = i32::MAX;
+        let mut max_rows = i32::MIN;
+        let mut max_columns = i32::MIN;
+
+        for position in entity_grid.keys() {
+            min_rows = min_rows.min(position.y);
+            min_columns = min_columns.min(position.x);
+            max_rows = max_rows.max(position.y);
+            max_columns = max_columns.max(position.x);
+        }
+
+        for (&position, &entity) in entity_grid.iter() {
+            let neighbors = self.neighbors.entry(entity).or_insert(NavNeighbors::EMPTY);
+
+            for direction in CompassOctant::VARIANTS {
+                // Don't overwrite existing connections
+                if neighbors.get(direction).is_some() {
+                    continue;
+                }
+
+                let maybe_neighbor = if should_loop {
+                    Self::find_nearest_grid_neighbor_in_direction::<true>(
+                        &entity_grid,
+                        min_rows,
+                        min_columns,
+                        max_rows,
+                        max_columns,
+                        direction,
+                        position,
+                    )
+                } else {
+                    Self::find_nearest_grid_neighbor_in_direction::<false>(
+                        &entity_grid,
+                        min_rows,
+                        min_columns,
+                        max_rows,
+                        max_columns,
+                        direction,
+                        position,
+                    )
+                };
+
+                if let Some(neighbor) = maybe_neighbor {
+                    // Entities should not be neighbors of themselves
+                    // Inserting the same entity at multiple locations is supported,
+                    // so we need to check for this case
+                    if neighbor != entity {
+                        // PERF: we could also add the reverse edge here, to save work later
```

Comment:
Multiple positions for the same node would make this very tricky/not possible.

Reply:
Right 🤔 Okay, I'll cut the PERF note.